test-data/unit/check-unreachable-code.test

```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
+    reveal_type(c)  # E: This branch is inferred to be unreachable
+
+d: int
+if False:
+    reveal_type(d)  # E: This branch is inferred to be unreachable
+
+e: int
+if True:
+    reveal_type(e)  # N: Revealed type is 'builtins.int'
+else:
+    reveal_type(e)  # E: This branch is inferred to be unreachable
+
+[builtins fixtures/isinstancelist.pyi]
+
+[case testUnreachableFlagTryBlocks]
+# flags: --disallow-inferred-unreachable
+
+def foo(x: int) -> int:
+    try:
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+        return x
+        reveal_type(x)      # E: This branch is inferred to be unreachable
+    finally:
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+        if True:
+            reveal_type(x)  # N: Revealed type is 'builtins.int'
+        else:
+            reveal_type(x)  # E: This branch is inferred to be unreachable
+
+def bar(x: int) -> int:
+    try:
+        if True:
+            raise Exception()
+        reveal_type(x)      # E: This branch is inferred to be unreachable
+    except:
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+        return x
+    else:
+        reveal_type(x)      # E: This branch is inferred to be unreachable
+
+def baz(x: int) -> int:
+    try:
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+    except:
+        # Mypy assumes all lines could throw an exception
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+        return x
+    else:
+        reveal_type(x)      # N: Revealed type is 'builtins.int'
+        return x
+[builtins fixtures/exception.pyi]
+
+[case testUnreachableFlagIgnoresSemanticAnalysisUnreachable]
+# flags: --disallow-inferred-unreachable --python-version 3.7 --platform win32 --always-false FOOBAR
+import sys
+from typing import TYPE_CHECKING
+
+x: int
+if TYPE_CHECKING:
+    reveal_type(x)  # N: Revealed type is 'builtins.int'
+
+if sys.platform == 'darwin':
+    reveal_type(x)
+
+if sys.version_info == (2, 7):
+    reveal_type(x)
+
+FOOBAR = ""
+if FOOBAR:
+    reveal_type(x)
+else:
+    reveal_type(x)  # N: Revealed type is 'builtins.int'
+[builtins fixtures/ops.pyi]
+
+[case testUnreachableFlagIgnoresSemanticAnalysisExprUnreachable]
+# flags: --disallow-inferred-unreachable --always-false FOOBAR
+import sys
+from typing import TYPE_CHECKING
+
+FOOBAR = ""
+
+def foo() -> bool: ...
+
+lst = [1, 2, 3]
+
+a = FOOBAR and foo()
+b = (not FOOBAR) or foo()
+c = 1 if FOOBAR else 2
+d = [x for x in lst if FOOBAR]
+[builtins fixtures/list.pyi]
+
+[case testUnreachableFlagOkWithDeadStatements]
+# flags: --disallow-inferred-unreachable
+from typing import NoReturn
+def assert_never(x: NoReturn) -> NoReturn:
+    assert False
+
+def nonthrowing_assert_never(x: NoReturn) -> None: ...
+
+def expect_str(x: str) -> str: pass
+
+x: int
+if False:
+    assert False
+    reveal_type(x)
+
+if False:
+    raise Exception()
+    reveal_type(x)
+
+if False:
+    assert_never(x)
```

Comment:
Random thought (no need to address in this PR): What do you think about providing a magic utility function such as `unreachable()` for easily silencing errors about unreachable code?

Reply:
That could be handy, though possibly annoying to use for people with linters (since they don't understand mypy magic functions). We could work around this by adding the function to `mypy_extensions`? Or maybe that's overkill?

Another idea: once the "tag all error messages with a code" PR is landed, the user could maybe add a `# type: ignore` for just the "Statement is unreachable" warning?

Reply:
Or another idea -- what if we instead modified mypy so that we *mandate* that any calls to `unreachable()` are actually unreachable?

We could also modify this function so that optionally take in any number of arguments: if the checker runs into this function, it'll run `reveal_type(...)` on the provided args before reporting an error that the statement was actually reachable. This would also let us supplant this "assert never" pattern (and providing a runtime implementation of the same logic in `mypy_extensions` or something would feel less like overkill).

Reply:
Yeah, that would be how I imagined it to work. Not sure about the arguments though. My idea would be to do something like this:

```py
X = Union[A, B]

def f(x: X) -> None:
    if isinstance(x, A):
        ...
    elif isinstance(x, B):
        ...
    else:
        unreachable()  # Error only if we forgot to handle some case
```

Maybe we could automatically recognize that pattern and print out the inferred type of `x` if we hit an error.